Integrating Chart.js in Django (#3255)
section-engineering-education/engineering-education: content/articles/integrating-chart.js-in-django/index.md

> Chart.js is an open-source HTML5 based javascript chart and renders in HTML5 canvas. It basically helps in visualizing data using javascript. 

Reviewer: In addition to this, can you talk about what other common libraries do we use for charts? How is this better than them? 

2 - 3 lines about this would be enough. This is just to make the intro more descriptive.

In this tutorial, we'll be looking at how to integrate Chart.js in Django by creating a simple project.

### Prerequisites
To follow along with this tutorial, you'll need [Python3](https://www.python.org/downloads/) installed on your machine.

> A little understanding of Django would help the reader to follow along better.

Reviewer: ```suggestion
A little understanding of Django would help the reader to follow along better. To understand more about Django, it is recommended to read these articles on [Django CRUD APIs](/engineering-education/django-crud-api/) and [Django API documentation](/engineering-education/django-api-documentation/).
```

### Table of contents
1. [Creating and setting up a new Django project](#creating-and-setting-up-a-new-django-project)
2. [Creating a model](#creating-a-model)
3. [Creating a view](#creating-a-view)
4. [Creating templates](#creating-templates)
5. [Conclusion](#conclusion)

### Creating and setting up a new Django project
Let's start by creating a new virtual environment.

A virtual environment allows you to create different spaces on your computer, with a different set of libraries and versions.

By creating a virtual environment, you'll be able to separate the necessary library installation for a project, without having to install them globally.

Now, create a virtual environment `env` by running the command below:

```bash

> $ python -m venv env

Reviewer: You could remove `$` from these code snippets.

This would allow the readers to copy-paste the codes easily

Reviewer: Do this for all the below code snippets

```

Here, we tell Python to create the virtual environment in a folder named `env` in the current directory.

On creation, activate the virtual environment using the following command:

```bash
$ source env/bin/activate
```

On activating the environment, install Django using the following command:

```bash
$ pip install django
```

Now, let's create a new Django project `Django-charts` using:

```bash
$ django-admin startproject twilio_proj
```

> Then navigate to the `Django-charts` directory and create a Django app.

Reviewer: ```suggestion
Then, navigate to the `Django-charts` directory and create a Django app:
```

```bash
$ cd Django-charts
$ django-admin startapp charts
```

Let's add our app to the list of installed apps.

> Navigate to the `Django-charts` directory and edit the `settings.py` file.

Reviewer: ```suggestion
Navigate to the `Django-charts` directory and edit the `settings.py` file:
```

```python
INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'editors',
]
```

> ### Creating a model

Reviewer: What's a model? Why do we need to create it?

> We'll be creating a chart that shows the number of users for different text editors. So, for our model, we'll be creating just two fields. User field and text editor field.

Reviewer: ```suggestion
We'll be creating a chart that visualize the trends in usage for different text editors.

Here, for our model, we'll be creating two fields - User field `num_users` and text editor field `editor_name`.
```

In the `editors` directory, edit the `models.py` file and add the following lines of code to it:

```python
from django.db import models

#defining a simple class
class Editors(models.Model):
    editor_name = models.CharField(max_length=200)
    num_users = models.IntegerField()

    def __str__(self):
        return "{}-{}".format(self.editor_name, self.num_users) 

```

Now let's register the model of Score in the `admin.py` file so that we can modify it in the Django admin section.

```python
from django.contrib import admin
from .models import Editors

admin.site.register(Editors)
```

The next step is to migrate our model to the database.

> Migrations are Django’s way of propagating changes you make to your models (adding a field, deleting a model, etc.) into your database schema.

Reviewer: Could you elaborate a little more on this?

Now, let's migrate our model to the database by running the commands below:

```bash
$ python manage.py makemigrations # migrating the app and database changes
$ python manage.py migrate        # final migrations
```

A superuser has the permissions to create, edit, update and delete data in Django admin. Create a superuser by running the command below:

```bash
$ python manage.py createsuperuser 
```

Now, we can log in to the admin page.

> To login to the admin section, serve the app again and go to http://localhost:8000/admin and log in with your superuser details.

Reviewer: ```suggestion
To login to the admin section, serve the app again and go to `http://localhost:8000/admin` and log in with your superuser details.
```

![Django admin login page](/engineering-education/content/articles/integrating-chart.js-in-django/login.png)

*Django admin login page*

![Django admin page](/engineering-education/content/articles/integrating-chart.js-in-django/admin.png)

*Django admin page*

Now, let's add an editor in the `Editors` category to test if its error-free.

Click on editor, then click `Add Editor` to add an editor.

![Add Editor](/engineering-education/content/articles/integrating-chart.js-in-django/add-editor.png)

*Add editor*

### Creating a view

> Now, let's create a view by adding the following to our `view.py` file:

Reviewer: What's a view? Why do you need to create it? Brief a little about it.

(Beginners would find it difficult to understand)

```python
from django.shortcuts import render
from django.views.generic import TemplateView
from .models import Editors

# Creating views
class EditorChartView(TemplateView):
    template_name = 'editors/chart.html'

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["qs"] = Editors.objects.all()
        return context
```

We do not have our `chart.html` file created yet. Let's create that by creating a templates folder in our `editors` directory. It should look like this `templates/editors/chart.html`.

### Creating templates
We start by creating a templates directory in our base directory. This templates directory will house our `base.html` file.

> base.html is basically a template that you extend on every page of your website.

Reviewer: ```suggestion
`base.html` is basically a template that you extend on every page of your website.
```

<script>
// jquery function
$(document).ready(function(){
        var ctx = document.getElementById('myChart').getContext('2d');
        var myChart = new Chart(ctx, {
            type: 'doughnut',
            data: {
                labels: [{%for data in qs%}'{{data.editor_name}}',{%endfor%}] //loop through queryset, 
                datasets: [{
                    label: '# of users',
                    data: [{%for data in qs%}{{data.num_users}},{%endfor%}],
                    backgroundColor: [
                        'rgba(255, 99, 132, 0.2)', 
                        'rgba(54, 162, 235, 0.2)',
                        'rgba(255, 206, 86, 0.2)',
                        'rgba(75, 192, 192, 0.2)',
                        'rgba(153, 102, 255, 0.2)',
                        'rgba(255, 159, 64, 0.2)'
                    ],
                    borderColor: [
                        'rgba(255, 99, 132, 1)',
                        'rgba(54, 162, 235, 1)',
                        'rgba(255, 206, 86, 1)',
                        'rgba(75, 192, 192, 1)',
                        'rgba(153, 102, 255, 1)',
                        'rgba(255, 159, 64, 1)'
                    ],

>                     borderWidth: 1

Reviewer: Could you please brief on what is being done here?

                }]
            },
            options: {
                scales: {
                    y: {
                        beginAtZero: true
                    }
                }
            }
});
    });
</script>
{%endblock scripts%}

{%block content%}
<!-- displaying the chart -->
<!-- you can also play around with the width and height to increase or decrease the chart size -->
<canvas id="myChart" width="400" height="100"></canvas>  
{%endblock content%}

```

> Next, we edit our urls.py by adding the following code:

Reviewer: ```suggestion
Next, we edit our `urls.py` by adding the following code:
```

```python
from django.contrib import admin
from django.urls import path
from editors.views import EditorChartView

urlpatterns = [
    path('admin/', admin.site.urls),
    path('', EditorChartView.as_view(), name='index')
]

```

Now, let's add a few text editors and a number of users to our Django admin.

![Editors](/engineering-education/content/articles/integrating-chart.js-in-django/editors.png)

*editors*

Now, let's test this out by starting the server with the following command:

```bash
$ python manage.py runserver 
```

![chart](/engineering-education/content/articles/integrating-chart.js-in-django/chart.png)

*it works*

### Conclusion
To conclude, we have learned about Chart.js. We have also learned to integrate charts in our Django app.

There are tons of things you can achieve with chart.js. In this tutorial we used a `doughnut` chart type, there other chart types you can use. You can also check the [Chart.js](https://www.chartjs.org/docs/latest/charts/) documentation for more info on chat types.

> 

Reviewer: Can you add a GitHub repo link that contains full code?
